src/third/redux-persist/createPersistor.js

```diff
@@ -56,9 +57,8 @@ export default function createPersistor (store, config) {
         storage.setItem(storageKey, serializer(substate)).catch(warnIfSetError(key))
       }
       writeInProgress = false
+      lastState = state
     })()
-
-    lastState = state
```

Comment:
createPersistor [nfc]: Include more code in an async IIFE.

There's actually a change happening here which is easy to miss in the source code, and it's not obvious whether it could affect visible behavior.

In the old code, we invoke the async IIFE, then update `lastState` and return. Then later when the whole JS stack is empty, the task created by that async IIFE gets run, and the code in its body starts executing. If the loop condition is true so it hits the `await` inside there, it creates a new task and returns, and later that task gets run so the rest of the loop body gets executed, and so on with further tasks until the loop condition is false so there's no more `await`.

In the new code, we invoke the (larger) async IIFE and return. The update to `lastState` doesn't happen immediately, not until some subsequent task: whichever one finds the loop condition is false.

I *think* this doesn't end up changing behavior, but only because I think `_resetLastState` can't end up getting called between when this code runs and when a write is complete, because I think we don't ever do a write until after the only time we'd call `_resetLastState`. But that fact is a bit nonlocal and delicate, and I'd have to look harder even to be sure it's true.

I think the right fix here is that the update `lastState = state` should really happen atomically with the point where we compare `state` to the old `lastState`:
```js
       Object.keys(state).forEach((key) => {
         if (!passWhitelistBlacklist(key)) return
         if (lastState[key] === state[key]) return
```

That would emphasize the point that it represents the last state we wrote out, and we're comparing the new state against the last one we wrote out, so that that comparison is just an optimization to skip writing out redundantly the parts that already have the latest version written out.

In the old code, that's happening because the only thing that comes between the comparison and the update is the async IIFE, which doesn't do anything immediately -- though that'd be clearer if they were next to each other in the source.

In the new code, if we just do the update right after the comparison loop, then I think things will be good.

Reply:
OK, as described below: https://github.com/zulip/zulip-mobile/pull/4694#discussion_r691735576 I see that updating `lastState` here at the end was actually in my draft, and I'm fine with putting it there. (As discussed in that thread there's a bit of a loophole in the error case where the write fails, but that's a lower-order concern than the problems this subsystem has anyway and that this PR mitigates, and we can handle that later.)

Should identify in the commit message that that's happening, though. And should also mention in… well, `_resetLastState` should never be called after a write may already be in progress, because if so it may not be effective. Should mention that in the comment on that method.

Reply:
> Should mention that in the comment on that method.

I've made that comment a jsdoc, since it seemed more of an interface question than implementation.